doc/xgboost_on_sqlflow.md

```diff
@@ -0,0 +1,139 @@
+
+Below is a demo about training/predicting via xgboost :
+
+```sql
+// sample clause of train
+select 
+    c1, c2, c3, c4, c5 as class
+from kaggle_credit_fraud_training_data
+TRAIN XGBoostEstimator
+WITH
+  booster = "gbtree"
+  objective = "logistic:binary"
+  eval_metric = "auc"
+  train_eval_ratio = 0.8
+COLUMN
+  NUMERIC(c1) as fc1,
+  BUCKETIZED(fc1, [0, 10, 100]) as fc2,
+  CROSS([fc1, fc2, c3]) as fc3,
+  c4 as fc4
+LABEL class
+INTO sqlflow_models.xgboost_model_table;
+
+// sample clause of predict
+select 
+    c1, c2, c3, c4
+from kaggle_credit_fraud_development_data
+PREDICT kaggle_credit_fraud_development_data.class
+USING sqlflow_models.xgboost_model_table
+COLUMN
+  NUMERIC(c1) as fc1,
+  BUCKETIZED(fc1, [0, 10, 100]) as fc2,
+  CROSS([fc1, fc2, c3]) as fc3,
+  c4 as fc4;
+```
+
+### Implementation
+
+As `codegen.go` generating _TensorFlow_ code from sqlflow AST,
+we will add `codegen_xgboost.go` which translate sqlflow AST into a python launcher program of _xgboost_. 
+
+In _ant-xgboost_, there exists an incubating module named [_xgblauncher_](https://github.com/alipay/ant-xgboost/tree/ant_master/xgboost-launcher), 
+an extendable, cloud-native xgboost based machine learning pipeline, 
+which provides a python API for building custom `DataSource` and `ModelSource`.
+
+Below is a demonstration of DataSource/ModelSource API.
+ 
+```python
+class DataSource:
+    """
+    DataSource API
+    A handler of data reading/writing, which is compatible with both single-machine and distributed runtime.
+    """
+    def __init__(self, 
+                 rank: int, 
+                 num_worker: int,
+                 column_conf: configs.ColumnFields,
+                 source_conf):
+        pass
+        
+    @abstractmethod
+    def read(self) -> Iterator[XGBoostRecord]:
+        pass
+
+    @abstractmethod
+    def write(self, result_iter: Iterator[XGBoostResult]):
+        pass
+
+    
+class ModelSource:
+    """
+    ModelSource API
+    A handler by which XGBLauncher save/load model(booster) and related information.
+    """
+    def __init__(self, source_conf):
+        pass
+
+    @abstractmethod
+    def read_buffer(self, model_path: str) -> bytes:
+        pass
+
+    @abstractmethod
+    def write_buffer(self, buf: bytes, model_path: str):
+        pass
+
+    @abstractmethod
+    def read_lines(self, model_path: str) -> List[str]:
+        pass
+
+    @abstractmethod
+    def write_lines(self, lines: List[str], model_path: str):
+        pass
+``` 
+
+
+With _xgblauncher_, we can launch _xgboost_ from sqlflow AST via a lightweight python `code template` and a corrsponding `filler`.
+The `code template` roughly includes  components as follows: 
+
+* `TFDataSource` that is responsible for fetching and pre-processing data via tf.feature_columns API.
+   Data will be fetched in mini-batch style by executing TF compute graph with mini-batch data feed by `sqlflow_submitter.db.db_generator`.
```

Comment:
We want to support as many `feature_columns` as sqlflow semantic does.
Although feeding xgboost with vectorized features like bucketized vectors is not always a good idea, xgboost can handle them anyway.  Since linear model is available in a XGBoost as an alternative to tree model,  I think it is reasonable to train a XGBoost model with vectorized features.

Reply:
Sure, you can find the list of currently supported feature column here: https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/user_guide.md#feature-columns.